crates/wasm-smith/src/core.rs

```diff
@@ -2813,15 +2836,18 @@ flags! {
     #[allow(missing_docs)]
     #[cfg_attr(feature = "_internal_cli", derive(serde_derive::Deserialize))]
     pub enum InstructionKind: u16 {
-        Numeric,
-        Vector,
-        Reference,
-        Parametric,
-        Variable,
-        Table,
-        Memory,
-        Control,
-        Aggregate,
+        NumericNonFloat = 1 << 0,
+        Numeric = (1 << 1) | (1 << 0),
```

Comment:
Since we don't have a way to express subsets or instructions, can we make these options be `NumericFloat` and `NumericNonFloat`? I think that would make it clearer what is required to enable the generation of certain instructions. Otherwise people might think that enabling `Numeric` was enough to get integer instructions, but if I understand correctly, it would only enable the float instructions and not the integer instructions.

Reply:
Another possible solution would be to not change `InstructionKind` here but instead add a new condition guarding all float-related instructions on `Config::disallow_floats` perhaps?

Reply:
> Otherwise people might think that enabling `Numeric` was enough to get integer instructions, but if I understand correctly, it would only enable the float instructions and not the integer instructions.

I might have overlooked something, but the intention is that `Numeric` allows all numeric instructions to be generated including both integers and floats. I tested a small example with `Numeric` as the only `allowed_instructions`, and it seems to work properly.

If the bits in the `FlagSet` are labeled `... b2 b1 b0`, then the flags are defined to set the following bits
- `Numeric`: `b1 b0`, `NumericNonFloat`: `b0`
- `Vector`: `b3 b2`, `VectorNonFloat`: `b2`
- `Memory`: `b9 b8`, `MemoryNonFloat`: `b8`
In each case then, setting the `Foo` flag also sets the `FooNonFloat` flag, and we have
```
assert!(InstructionKinds::new(&[InstructionKind::Numeric]).contains(InstructionKind::NumericNonFloat));
assert!(InstructionKinds::new(&[InstructionKind::Vector]).contains(InstructionKind::VectorNonFloat));
assert!(InstructionKinds::new(&[InstructionKind::Memory]).contains(InstructionKind::MemoryNonFloat));
```

Reply:
`Integral` instead of `NonFloat` might be a better name here.

Reply:
I think the overlapping-bits representation here would probably work (in contrast to @fitzgen's observation where `Numeric` as-written should generate integer instructions). As for the original intention here, the hope was to basically help fuzzing generate interesting modules by only giving it access to certain classes of instructions. In that sense the exact set of classes here and the exact definitions don't matter too too much.

Personally I'd advocate for not changing `InstructionKind` though, but only for the reason that I think it could be confusing to have `allow_floats: false` and then have float instructions generated because you might have also forgotten to disable float-related instruction kinds here. Or, @Scott-Guest, do you have a use case for disabling the f32/f64 type but enabling float-related instructions? If not then I'd personally prefer that `allow_floats: false` additionally disables all float-related instructions without having to additionally modify `InstructionKind` classes.

If we keep the `InstructionKind` changes, though, then I'd agree that we should probably split it into `Numeric{Int,Float}` or similar names.

Reply:
> Or, @Scott-Guest, do you have a use case for disabling the f32/f64 type but enabling float-related instructions?

Nope - only use case I have is fully disabling anything related to floats. Changing the `InstructionKinds` was just an easy way to implement this.

> I think it could be confusing to have allow_floats: false and then have float instructions generated because you might have also forgotten to disable float-related instruction kinds here.

We won't generate float instructions if `allow_floats: false` regardless of the instruction kinds - see the other comment I tagged you in. 

> If we keep the InstructionKind changes, though, then I'd agree that we should probably split it into Numeric{Int,Float} or similar names.

I agree, but this is starting to feel pretty messy and non-orthogonal. There are really two concerns here:
- The general categories of instructions we generate
- The types of values that we generate

Let me try implementing a `allowed_value_types: ValTypes` config option instead - I'll open a separate PR.

Reply:
Oh it's ok to not go to too much effort I think, you raise a good point that the `InstructionKind` enum here is already a pretty handy way to implement this today and works relatively well. Given that I'm pretty convinced that this is reasonable. Maybe modulo a small rename from `NumericNonFloat` to `NumericInt` perhaps? 

Otherwise though if @fitzgen agrees I think it'd be reasonable to merge.

Reply:
SGTM :+1:

Reply:
Sounds good, I renamed `NonFloat` to `Int`. Feel free to merge!